### PR TITLE
context processors

### PR DESCRIPTION
**Pull request by imbolc**
Added tests, docs and 'aiohttp_jinja2_' prefixes

### CODE REVIEW

File: aiohttp_jinja2/__init__.py [hunk 1]
```diff
@@ -49,6 +51,10 @@ def render_string(template_name, request, context, *, app_key):
         text = "context should be mapping, not {}".format(type(context))
         # same reason as above
         raise web.HTTPInternalServerError(reason=text, text=text)
+    if REQUEST_CONTEXT_KEY in request:
+        for k, v in request.get(REQUEST_CONTEXT_KEY, {}).items():
+            if k not in context:
```
Comment (asvetlov) [inline aiohttp_jinja2/__init__.py]: I think we should be very explicit here by raising exception if context keys overlap each other.

File: aiohttp_jinja2/__init__.py [hunk 2]
```diff
@@ -81,3 +87,19 @@ def wrapped(*args):
...
     return wrapper
+
+
+@asyncio.coroutine
+def context_processors_middleware(app, handler):
+    def middleware(request):
+        request[REQUEST_CONTEXT_KEY] = {}
+        for processor in app[APP_CONTEXT_PROCESSORS_KEY]:
+            request[REQUEST_CONTEXT_KEY].update(
+                (yield from processor(request)))
```
Comment (asvetlov) [inline aiohttp_jinja2/__init__.py]: If you do care about overlapping keys in `render_string` you should care here also.

**Comment by imbolc** [comment]
Do you mean overlapping between processors or with handler context too?
And I think in some cases overlapping could be useful, e.g.

```
processors = [
    auth_with_cookie,
    auth_with_token,
    auth_with_ip,
]
```

Where each processor doing somethin like:

```
def auth_with_...(request):
    context = request['context']
    if not context.get('user'):
        return {'user': do_auth()}  # returns user on None
    return {}



```

**Comment by asvetlov** [comment]
Looks like I've got your intention.

Your code is following last-win strategy actually: last context processor wins but explicit context in `render_string` overrides data of context processors.

That's fine but the behavior should be documented.

Also I suggest changing `render_string` code to 
```
    if REQUEST_CONTEXT_KEY in request:
        cxt = request.get(REQUEST_CONTEXT_KEY)  # it's not None now, middleware guarantee this
        ctx.update(context)
        context = ctx
```

I believe it describes your intention explicitly.

**Comment by imbolc** [comment]
I used this way in the first iteration:
https://github.com/imbolc/aiohttp_jinja2/commit/26b9ace3b151d046b48925e169d1e83483910a0a#diff-3b27daa327d57f529f99a1ab813656c3R41

But with `copy`, because it's possible to render more than one template in
a handler and we can't change `request['context']` in this case. Do you
think this way more explicitly?

**Comment by asvetlov** [comment]
Got it. 
LGTM.
I'm ready to merge if you have no updates.

**Comment by imbolc** [comment]
Yes, i've updated the docs. Thanks :)

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**